dawn/src/dawn/Optimizer/PassMultiStageMerger.cpp

```diff
@@ -0,0 +1,168 @@
+  multiStageDependencyGraph.merge(*doMethod.getDependencyGraph());
+
+  // Try all possible loop orders while *favoring* a parallel loop order. Note that a parallel loop
+  // order can be changed to forward or backward.
+  //
+  //                 MULTI-STAGE
+  //
+  //             |  P  |  F  |  B  |           P = Parallel
+  //        -----+-----+-----+-----+           F = Foward
+  //    S     P  | PFB    F     B  |           B = Backward
+  //    T   -----+                 +           X = Incompatible
+  //    A     F  |  F     F     X  |
+  //    G   -----+                 +
+  //    E     B  |  B     X     B  |
+  //        -----+-----------------+
+  //
+  std::vector<iir::LoopOrderKind> possibleLoopOrders;
+
+  if(multiStageLoopOrder == iir::LoopOrderKind::Parallel &&
+     stageLoopOrder == iir::LoopOrderKind::Parallel)
+    possibleLoopOrders = {iir::LoopOrderKind::Parallel, iir::LoopOrderKind::Forward,
+                          iir::LoopOrderKind::Backward};
+  else if(stageLoopOrder == iir::LoopOrderKind::Parallel)
+    possibleLoopOrders.push_back(multiStageLoopOrder);
+  else
+    possibleLoopOrders.push_back(stageLoopOrder);
+
+  if(multiStageDependencyGraph.empty())
+    return ReturnType(multiStageDependencyGraph, possibleLoopOrders.front());
+
+  // If the resulting graph isn't a DAG anymore that isn't gonna work
+  if(!multiStageDependencyGraph.isDAG())
+    return ReturnType(std::nullopt, multiStageLoopOrder);
+
+  // Check all possible loop orders if there aren't any vertical conflicts
+  for(auto loopOrder : possibleLoopOrders) {
+    auto conflict = hasVerticalReadBeforeWriteConflict(multiStageDependencyGraph, loopOrder);
+    if(!conflict.CounterLoopOrderConflict)
+      return ReturnType(multiStageDependencyGraph, loopOrder);
+  }
+
+  return ReturnType(std::nullopt, multiStageLoopOrder);
+}
+
+/// @brief Check if we can merge the stage into the multi-stage, possibly changing the loop order.
+/// @returns the the enew dependency graphs of the multi-stage (or NULL) and the new loop order
+template <typename ReturnType =
+              std::pair<std::optional<iir::DependencyGraphAccesses>, iir::LoopOrderKind>>
+ReturnType isMergable(const iir::MultiStage& thisMS, const iir::MultiStage& otherMS) {
+  iir::LoopOrderKind thisLoopOrder = thisMS.getLoopOrder();
+  for(const auto& thisStage : thisMS.getChildren()) {
+    auto dependencyGraphLoopOrderPair = isMergable(*thisStage, thisLoopOrder, otherMS);
+    if(dependencyGraphLoopOrderPair.first) {
+      return dependencyGraphLoopOrderPair;
+    }
+  }
+  return ReturnType(std::nullopt, thisLoopOrder);
+}
+
+PassMultiStageMerger::PassMultiStageMerger(OptimizerContext& context)
+    : Pass(context, "PassMultiStageMerger") {
+  dependencies_.push_back("PassSetStageGraph");
+}
+
+bool PassMultiStageMerger::run(const std::shared_ptr<iir::StencilInstantiation>& instantiation) {
+  // Do we need to run this Pass?
+  bool doMultiStageMerge = false;
```

Comment:
Consider replacing with `if (std::any_of(instantiation->getStencil().cbegin(), ..., [](const std::unique_ptr<Stencil>& stencil) { return stencil->getChildren().size() > 1; }) { return true; }`. I like this more, but maybe not everyone does...